dylint/src/package_options/revs.rs

```diff
@@ -1,123 +1,276 @@
-    curr_rev: Option<Rev>,
+    repo: Rc<Repository>,
+    quiet: bool,
 }
 
 impl Revs {
     pub fn new(quiet: bool) -> Result<Self> {
-        let repository = clippy_repository(quiet)?;
-
-        Ok(Self { repository })
+        let start = Instant::now();
+        let repo = clippy_repository(quiet)?;
+        if !quiet {
+            eprintln!("Revs repository initialization took: {:?}", start.elapsed());
+        }
+        Ok(Self { repo, quiet })
     }
 
-    #[allow(clippy::iter_not_returning_iterator)]
-    pub fn iter(&self) -> Result<RevIter> {
-        let path = self
-            .repository
-            .workdir()
-            .ok_or_else(|| anyhow!("Could not get working directory"))?;
-        let object = {
-            let head = self.repository.head()?;
-            let oid = head.target().ok_or_else(|| anyhow!("Could not get HEAD"))?;
-            self.repository.find_object(oid, Some(ObjectType::Commit))?
-        };
-        let commit = object
-            .as_commit()
-            .ok_or_else(|| anyhow!("Object is not a commit"))?;
-        let version = clippy_utils_package_version(path)?;
-        let channel = toolchain_channel(path)?;
+    // Extracts Rev info from a specific commit by parsing blobs.
+    fn get_rev_info_for_commit(&self, commit: &Commit) -> Result<Option<Rev>> {
+        let tree = commit.tree().context("Failed to get commit tree")?;
         let oid = commit.id();
-        Ok(RevIter {
-            repository: self.repository.clone(),
-            commit: commit.clone(),
-            curr_rev: Some(Rev {
-                version,
-                channel,
-                oid,
-            }),
-        })
+
+        // Try to get version from Cargo.toml
+        let version = if_chain! {
+            if let Ok(cargo_entry) = tree.get_path(Path::new("Cargo.toml"));
+            if let Ok(blob) = self.repo.find_blob(cargo_entry.id());
+            if let Ok(content) = std::str::from_utf8(blob.content());
+            if let Ok(cargo_toml) = content.parse::<Value>();
+            if let Some(package) = cargo_toml.get("package");
+            if let Some(version_val) = package.get("version");
+            if let Some(version_str) = version_val.as_str();
+            then {
+                Some(version_str.to_string())
+            } else {
+                None
+            }
+        };
+
+        // Return early if version was not found
+        let Some(version_str) = version else {
+            return Ok(None);
+        };
+
+        // Try to get channel from rust-toolchain
+        let channel = if_chain! {
+            if let Ok(toolchain_entry) = tree.get_path(Path::new("rust-toolchain"));
+            if let Ok(blob) = self.repo.find_blob(toolchain_entry.id());
+            if let Ok(content) = std::str::from_utf8(blob.content());
+            then {
+                content.trim().to_string()
+            } else {
+                return Ok(None);
+            }
+        };
```

Comment:
I think the problem is here. It looks like this code is reading the entire `rust-toolchain` file and treating it as the toolchain channel. I think `rust-toolchain` files used to work that way, but no longer. I recommend using the `toolchain_channel` function to get the toolchain channel.